ast/Trees.h

```diff
@@ -73,21 +133,28 @@ class ParsedFilesOrCancelled final {
     std::vector<ParsedFile> &result();
 };
 
-template <class To> To *cast_tree(Expression *what) {
-    static_assert(!std::is_pointer<To>::value, "To has to be a pointer");
-    static_assert(std::is_assignable<Expression *&, To *>::value, "Ill Formed To, has to be a subclass of Expression");
-    return fast_cast<Expression, To>(what);
+template <class To> To *cast_tree(TreePtr &what) {
+    return fast_cast<Expression, To>(what.get());
 }
 
 // A variant of cast_tree that preserves the const-ness (if const in, then const out)
-template <class To> const To *cast_tree_const(const Expression *what) {
-    static_assert(!std::is_pointer<To>::value, "To has to be a pointer");
-    static_assert(std::is_assignable<Expression *&, To *>::value, "Ill Formed To, has to be a subclass of Expression");
-    return fast_cast<Expression, To>(const_cast<Expression *>(what));
+template <class To> To const *cast_tree_const(const TreePtr &what) {
+    return fast_cast<Expression, To>(const_cast<Expression *>(what.get()));
+}
```

Comment:
@DarkDimius what do you think about making `cast_tree_const` just a specialization of `const_tree` that preserves  const-ness instead of having a completely different function?

Reply:
happy to experiment, but let's do it separately.
Having them separate seems benefitial to as it makes us more explicit around semantics that you _wanted_ a const one.

Sometimes you want a const pointer out of non-cost one, and, unless you explicitly do instantiation, you won't get it in merged version.